source/common/grpc/context_impl.cc

```diff
@@ -49,7 +49,9 @@ void ContextImpl::chargeStat(const Upstream::ClusterInfo& cluster, Protocol prot
       symbol_table_.join({protocolStatName(protocol), request_names.service_, request_names.method_,
                           status_stat_name});
 
-  cluster.statsScope().counterFromStatName(Stats::StatName(stat_name_storage.get())).inc();
+  cluster.statsScope()
+      .counterFromStatName(Stats::StatName(stat_name_storage.get()), absl::nullopt)
```

Comment:
this does make the PR a lot bigger (and was probably annoying to do; sorry about that). On the plus side, having only a single implementation in the various stores is a big win.

I think I agree with the rationale in the Google Style Guide pre-C++11.

https://google.github.io/styleguide/cppguide.html#Default_Arguments

However, with C++11 I feel that the argument against default-args on virtual functions misses the distinction between PURE interfaces and overrides, which are more explicit in C++11. i.e. if we define the default value only at the interface, and disallow changing the default for overrides, I don't see how what scenario doesn't behave correctly.

Moreover there is precedent in the Envoy code base for this pattern of default-args in interfaces. A simple grep finds:
```
./envoy/json/json_object.h:80:  virtual ObjectSharedPtr getObject(const std::string& name, bool allow_empty = false) const PURE;
./envoy/json/json_object.h:107:                                                      bool allow_empty = false) const PURE;
./envoy/json/json_object.h:132:                                                  bool allow_empty = false) const PURE;
./envoy/common/scope_tracker.h:26:  virtual void dumpState(std::ostream& os, int indent_level = 0) const PURE;
./envoy/common/pure.h:7:#define PURE = 0
./envoy/stream_info/filter_state.h:78:                       StateType state_type, LifeSpan life_span = LifeSpan::FilterChain) PURE;
./envoy/event/timer.h:43:                           const ScopeTrackedObject* object = nullptr) PURE;
./envoy/event/timer.h:53:                             const ScopeTrackedObject* object = nullptr) PURE;
./envoy/http/api_listener.h:27:                                    bool is_internally_created = false) PURE;
./envoy/http/header_map.h:602:  virtual void dumpState(std::ostream& os, int indent_level = 0) const PURE;
./envoy/http/codec.h:489:                                    bool is_internally_created = false) PURE;
```

WDYT?

The other alternative is to just use a different interface name for the methods that take tag-vectors.

I'm also OK with what you have if you strongly prefer but I'd be interested in @mattklein123 's opinion on this. There are other cases where we deviate from the Google style guide, and document that fact in https://github.com/envoyproxy/envoy/blob/master/STYLE.md .

Reply:
I'm fine with allowing default values on interface functions. I honestly don't know the history of why they are generically blocked.

Reply:
I found out why this is disallowed; it doesn't work as I thought it ought to:
```
struct Base {
  virtual void f(int i = 42) = 0;
};

struct Derived : Base {
  void f(int) override;
};

int main() {
  Derived d;
  Base& b = d;  // Use the same object via a different name and static type.
  b.f();  // OK, equivalent to d.Derived::f(42).
  d.f();  // Error.  Derived::f requires one argument.
}
```
I think we either have to explicitly give the nullopt arg at every counterFromStatName() call, or we'll have to use a non-virtual helper. I'm OK either way.

Reply:
The problem I ran into with the non-virtual helper is that `StatName` isn't a complete type at this point so I can't actually define the non-virtual helper on the top level `Scope` class. We could define `StatName` in include/envoy/common/stats/symbol_table.h, but that puts a fair bit of implementation logic into the include/ which seems undesirable. The other way to make this possible would be to use a virtual interface class for it, but I'm guessing we don't want that for the sake of keeping `StatName` simple?

Using a default value on the top level definition would be helpful for most of the use cases   since most things are invoking the functions from a `Scope` reference, so it might still be useful even if the default value overload doesn't inherit? It might end up being confusing, but if we're just trying to solve for common case it does make the API a bit simpler. 

That said I'm happy to keep the explicit value if we think that's the way to go.

Reply:
Can you define the non-virtual helper if it's passing through `const StatName&`  rather than `StatName`?

Reply:
I was able to get it working by using `const StatName&` and `StatNameTagVectorOptRef` everywhere, ptal